pallets/thea/src/lib.rs

```diff
@@ -0,0 +1,534 @@
+	}
+
+	// Hooks for Thea Pallet are defined here
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_n: BlockNumberFor<T>, mut remaining_weight: Weight) -> Weight {
+			// TODO: Calculate proper weight for single claim call on on_idle
+			let single_claim_weight: Weight = 100_000_000;
+
+			if remaining_weight < single_claim_weight {
+				// We need enough weight for at least one claim process if not it's a no-op
+				return remaining_weight
+			}
+
+			let mut accounts = <AccountWithPendingDeposits<T>>::get();
+			if accounts.is_empty() {
+				return remaining_weight
+			}
+
+			while let Some(account) = accounts.pop_first() {
+				if let Some(mut pending_deposits) = <ApprovedDeposits<T>>::get(&account) {
+					// FIXME: This leads to an infinite loop if execute_deposit fails
+					while let Some(deposit) = pending_deposits.pop() {
+						if let Err(err) = Self::execute_deposit(deposit, &account) {
+							// Force push is fine as it was part of the bounded vec
+							pending_deposits.force_push(deposit);
+							// We can't do much here other than to log an error.
+							log::error!(target:"runtime::thea::on_idle","Error while claiming deposit on idle: user: {:?}, Err: {:?}",account,err);
+						}
+						// reduce the remaining_weight
+						remaining_weight = remaining_weight.saturating_sub(single_claim_weight);
+						if remaining_weight.is_zero() {
+							break
+						}
+					}
+
+					if !pending_deposits.is_empty() {
+						<ApprovedDeposits<T>>::insert(&account, pending_deposits);
+						accounts.insert(account);
+					}
+				}
+			}
+
+			<AccountWithPendingDeposits<T>>::put(accounts);
+			remaining_weight
+		}
+	}
+
+	// Extrinsics for Thea Pallet are defined here
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// A Deposit transaction is called by the nodes with an aggregated BLS Signature
+		///
+		/// # Parameters
+		///
+		/// * `origin`: Active relayer
+		/// * `network_id`: id of the foreign chain network
+		/// * `bit_map`: The bit map of current relayer set that have signed the Deposit Transaction
+		/// * `recipient`: The address of the user who initiated a deposit
+		/// * `tx_hash`: Hash of the transaction on the foreign chain
+		/// * `asset_id`: The asset id of the asset being deposited
+		/// * `amount`: The amount of assets that have been deposited in foreign chain
+		/// * `bls_signature`: The aggregated signature of majority of relayers in current active
+		///   relayer set
+		#[pallet::weight(1000)]
+		pub fn approve_deposit(
+			origin: OriginFor<T>,
+			bit_map: u128,
+			bls_signature: [u8; 96],
+			payload: Payload<T::AccountId>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			ensure!(payload.amount > 0, Error::<T>::AmountCannotBeZero);
+			// Fetch Deposit Nonce
+			let nonce = <DepositNonce<T>>::get(payload.network_id.saturated_into::<Network>());
+			ensure!(payload.deposit_nonce == nonce + 1, Error::<T>::DepositNonceError);
```

Comment:
Sure, this selected block could be moved lower by execution chain since `nonce` required only on line number 306. So by moving 281-283 lines after 303 - other assertions would be executed first and calculation time in invalid scenario will be faster.

Reply:
Other assertions are computationally more expensive, So isn't it better to have it here?

Reply:
Ok